### PR TITLE
Design frontend

### PR DESCRIPTION
**Pull request by zmbq**
This pull request contains the fixes to the issues raised by users.

It also contains the first SignalR notification handler.

### CODE REVIEW

File: src/state/signalr-connection.ts [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+        this._currentHandler = handler;
+        if (this._utils) {
+            this._utils!.connectNotificationHandler(handler);
+        }
+    }
+
+    public unregisterNotificationHandler() {
+        if (this._currentHandler && this._utils) {
+            this._utils.disconnectNotificationHandler(this._currentHandler);
+        }
```
Comment (Bronson-Brown-deVost) [inline src/state/signalr-connection.ts]: I was a little confused about two things here.  Firstly the usage of this._currentHandler implies that there will only be one handler at a time running. Certainly we will need more than one sometimes.  Secondly, I think with the this._currentHandler variable you were rightly trying to make sure that you are calling "off" with a reference to the same function that "on" was called with (which you must do).  So I guess we probably need here some sort of array or object to hold all the currently registered NotificationHandlers.  You probably need this so that you can reregister the listeners on reconnect (check the docs for the new policy on autoreconnect for the v3.x signalr javascript client: https://docs.microsoft.com/en-us/aspnet/core/signalr/javascript-client?view=aspnetcore-3.1#reconnect-clients).

File: src/state/signalr-connection.ts [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+        await this.disconnect();
+        await this.connect();
+        if (subscribed) {
+            await this.subscribeEdition(subscribed);
+        }
+    }
+
+    public registerNotificationHandler(handler: NotificationHandler) {
+        this.unregisterNotificationHandler();
+        this._currentHandler = handler;
```
Comment (Bronson-Brown-deVost) [inline src/state/signalr-connection.ts]: I guess you would generally want to first unregister from this._currentHandler if it already exists.  I think you will also want to keep a record of which listeners have been engaged so that you can reregister them in cases where the connection has been lost and you have reconnected ( see the new instruction for auto reconnect: https://docs.microsoft.com/en-us/aspnet/core/signalr/javascript-client?view=aspnetcore-3.1#reconnect-clients).
Another possible problem I see is that if you register for, e.g., connectNotificationHandler with this method, then you register for some other listener in another method with a new NotificationHandler, your `this._utils.disconnectNotificationHandler(this._currentHandler)` below will fail, since this._currentHandler has been redefined and `this._currentHandler.handleUpdatedEdition` no longer refers to the `handler.handleUpdatedEdition` with which the "on" listener was created, so the listener will not be unregistered (the "on" and "off" must reference the same function, not just a function that does the same thing).

File: src/state/signalr-connection.ts [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+        this._status = 'connecting';
+        try {
+            await this._connection.start();
+            this._status = 'connected';
+            this._connection!.onclose(this.onConnectionClosed);
+
+            if (this._currentHandler) {
+                this._utils.connectNotificationHandler(this._currentHandler);
+            }
+            console.debug('SignalR connection opened, status is ', this._status);
```
Comment (Bronson-Brown-deVost) [inline src/state/signalr-connection.ts]: probably implement autoreconnect and add your onReconnecting and onReconnected functions here.

File: src/state/signalr-connection.ts (hunk 2 above)
Comment (zmbq) [reply src/state/signalr-connection.ts]: This is the older SignalR code. Look at the signalr branch for the new code.

I did not implement auto reconnection yet, I need to figure out how to do that, or even if there's any point in implementing it at all. If the connection is down and then up again, we may miss notifications in the meantime, which means we need to refresh whatever the user is currently viewing. I can add a mechanism to keep track of this (I need to somehow mark certain prepare functions as those that needs that need to be recalled after reconnecting), or we can just show a "Please refresh this page" notification to the user, much like Facebook does.

I suggest we try with the refresh message, which is very easy to implement, and add automatic recovery only if it bothers users in the real world.

**Comment by zmbq** [comment]
For some reason I can’t see this comment on github.

 

  I’m not sure whether I want one or multiple notification handles. I am certain I will only have one handler per notification function. I do not want multiple handlers for UpdatedEdition, for instance – just one that updates the state. Everybody else consults the state. If I need to add another form of notification (in NewArtefact, for instance, the component binds to an array that is not part of the state, which may be solved with a @watch or may not), I will use a non-signalR related event bus in the frontend.

 

  Itay.

**Comment by zmbq** [comment]
Tests seem to be unstable for now, yesterday they passed – on the exact same commit I used for the pull request. This has been an ongoing issue. I’m starting to wonder whether Cypress is the correct framework.

 

  The backend should most definitely reject floats if it expects integers. It should also reject negative numbers if it expects unsigned numbers. I don’t think there’s a way to force a number to be an integer in Typescript. I suspect these are bugs we’ll need to catch and handle ourselves. We’ll fix it once you send us a list.

 

 I’m not sure how to improve the opacity issue – when two images are at maximum, I want them both to be seen, at the same intensity. I’m not sure how do that.

File: src/state/signalr-connection.ts (hunk 2 above)
Comment (Bronson-Brown-deVost) [reply src/state/signalr-connection.ts]: Sorry, I just took a look at the other branch and that makes sense now.
I agree with you about reconnect, it is probably too fragile to do anything other than what you suggest.

**Comment by Bronson-Brown-deVost** [comment]
> I am certain I will only have one handler per notification function.

That is a very good policy!

> This has been an ongoing issue. I’m starting to wonder whether Cypress is the correct framework.

No worries, let give it a little more time before we give up.  We can send non-passing code through for now.

> I’m not sure how to improve the opacity issue – when two images are at maximum, I want them both to be seen, at the same intensity. I’m not sure how do that.

Ah, now I think I understand what you want, and it is a good idea.  So here are a few examples of how the opacity could match the slider settings in an intuitive way (I will let you figure out the implementation):
Example 1: two images selected, both sliders at full. the bottom image should have an opacity of 1, and the top image should have an opacity of 50% (that is your "both seen at the same intensity).
Example 2: two images selected, first slider full, second slider 50%.  Bottom image with opacity of 1, top image with opacity of .25.
Example 3: three images selected, all at maximum. The bottom image is at opacity 1, I think the top two are each at opacity 0.33.
Example 4: two images selected, first slider at 75%, second slider at 50%.  What to do?  I guess bottom image still at opacity of 1, the top image maybe at opacity .37 (=50% of 75%)?
Basically there should probably always be an image at opacity 1, there is no real use case for anything else.  Two overlaid images at .5 opacity just gives a "washed out" image, that is not likely to be what the user intended.

> The backend should most definitely reject floats if it expects integers. It should also reject negative numbers if it expects unsigned numbers. 

Ok. That makes sense.  I will dig those out right now.

**Comment by Bronson-Brown-deVost** [comment]
Ok.  So in the TranslateDTO, both parameters, x and y, are uint (32 bit unsigned).  I think that is the only one where I had real trouble.  

Too bad typescript is not set up to fix such issues, but alas it is just a bolt on checker/transpiler and `... as T` doesn't actually do any type conversion in the generated Javascript.  Fortunately, the API gives very clear error responses, so it doesn't take but a moment to see the problem.  I wonder if we can somehow transmit the underlying C# type in the comments to the typescript DTO or something like that, just for convenience sake, since we apparently can't do anything more typesafe.

**Comment by zmbq** [comment]
I think I see the logic. We want the *visibility* to be the same if both images have sliders at the same level. The visibility of the second image is its opacity multiplied by the (1 – the opacity of the first image). The visibility of the third image is its opacity multiplied by (1 – the visibility of the second image). The visibility of the first image is the opacity of the first image. An image with that isn’t shown has its opacity set to zero, and so its visibility is also zero.

 

  If that makes sense to you, I’ll work up the set of equations here and figure out how to solve them. We have the requested visibilities, we need to solve and find the opacity. So solving should be done top to bottom.

 

  Itay.

**Comment by zmbq** [comment]
OK, I fixed the specific bug, not the big problem of ints in Typescript (looking around, there is no solution). I wonder, though, how we got floats in the first place. They come from the bounding box of the polygon. Perhaps svg lets you drop on subpixel levels.

 

  Itay.

**Comment by Bronson-Brown-deVost** [comment]
> (1 – the opacity of the first image). The visibility of the third image is its opacity multiplied by (1 – the visibility of the second image). The visibility of the first image is the opacity of the first image. An image with that isn’t shown has its opacity set to zero, and so its visibility is also zero.

I guess we will have to try it out.  My febrous mind suggests we do the following: determine how many images are visible, let's say 3 for example.  The image that appears at the top of the <svg> element always has an opacity of 1 no matter what.  The next two images set their opacity with the following formula ((1 / numOfImages) * currentImageOpacitySetting) + ((1 - opacitySettingOfFirstImage) / (numOfImages - 1)).  So for example the first image is set to .75, the middle to .5 and the last to 0.25: the final opacity for the first image should nevertheless be 1 in the SVG element, the second image is ((1/3)*.5) + ((1 - .75) / (3 - 1)) = 0.29 and the third image is: ((1/3)*.25) + ((1 - .75) / (3 - 1)) = 0.20.  Maybe that will look good, I don't really know.  I probably won't be back again today, I have been very ill the last few days, but am starting to feel better.  Feel free to try out whatever you think might work well.

**Comment by zmbq** [comment]
Most of these have been addressed on the signalR branch, I'll open a new pull request.
